File: examples/c/ex_log.c
/*-
 * Public Domain 2008-2014 WiredTiger, Inc.
 *
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * ex_log.c
 * 	demonstrates how to logging and log cursors.
 */
#include <assert.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <wiredtiger.h>

const char *home = "./WT_EXLOG";
const char *home2 = "./WT_EXLOG2";
const char *uri = "table:logtest";

#define	CONN_CONFIG "create,cache_size=100MB,log=(archive=false,enabled=true)"
#define	MAX_KEYS	10

#define	RECORD_PRINT							\
    printf("LSN [%d][%" PRIu64 "].%d: "					\
	" record type %d optype %d txnid %" PRIu64 " fileid %d",	\
	lsn.file, lsn.offset, opcount, rectype,				\
	optype, txnid, fileid);						\
    if (logrec_key.size != 0)						\
	printf(" key size %" PRIu64, (uint64_t)logrec_key.size);	\
    if (logrec_value.size != 0)						\
	printf(" value size %" PRIu64, (uint64_t)logrec_value.size);	\
    printf("\n");

static int
setup_copy(WT_CONNECTION **wt_connp, WT_SESSION **sessionp)
{
	int ret;

	if ((ret = wiredtiger_open(home2, NULL,
	    CONN_CONFIG, wt_connp)) != 0) {
		fprintf(stderr, "Error connecting to %s: %s\n",
		    home, wiredtiger_strerror(ret));
		return (ret);
	}

	ret = (*wt_connp)->open_session(*wt_connp, NULL, NULL, sessionp);
	ret = (*sessionp)->create(*sessionp, uri,
	    "key_format=S,value_format=S");
	return (ret);
}

static int
compare_tables(WT_SESSION *session, WT_SESSION *sess_copy)
{
	WT_CURSOR *cursor, *curs_copy;
	int ret, ret_copy;
	const char *key, *key_copy, *value, *value_copy;

	ret = session->open_cursor(session, uri, NULL, NULL, &cursor);
	ret = sess_copy->open_cursor(sess_copy, uri, NULL, NULL, &curs_copy);

	while ((ret = cursor->next(cursor)) == 0) {
		ret_copy = curs_copy->next(curs_copy);
		ret = cursor->get_key(cursor, &key);
		ret = cursor->get_value(cursor, &value);
		ret_copy = curs_copy->get_key(curs_copy, &key_copy);
		ret_copy = curs_copy->get_value(curs_copy, &value_copy);
		if (strcmp(key, key_copy) != 0 ||
		    strcmp(value, value_copy) != 0) {
			fprintf(stderr,
			    "Mismatched: key %s, key_copy %s "
			    "value %s value_copy %s\n",
			    key, key_copy, value, value_copy);
			return (1);
		}
	}
	/*
	 * When the first cursor is done, the copy better be done too.
	 */
	ret_copy = curs_copy->next(curs_copy);
	assert(ret_copy != 0);
	cursor->close(cursor);
	curs_copy->close(curs_copy);
	return (0);
}

static int
walk_log(WT_SESSION *session)
{
	WT_CONNECTION *wt_conn2;
	WT_CURSOR *cursor, *cursor2;
	WT_LSN lsn, lsnsave;
	WT_ITEM logrec_key, logrec_value;
	WT_SESSION *session2;
	uint64_t txnid;
	uint32_t fileid, opcount, optype, rectype;
	int first, i, in_txn, ret;

	ret = setup_copy(&wt_conn2, &session2);
	/*! [log cursor open] */
	ret = session->open_cursor(session, "log:", NULL, NULL, &cursor);
	/*! [log cursor open] */
	ret = session2->open_cursor(session2, uri, NULL, "raw=true", &cursor2);
	i = 0;
	in_txn = 0;
	txnid = 0;
	memset(&lsnsave, 0, sizeof(lsnsave));
	while ((ret = cursor->next(cursor)) == 0) {
		/*! [log cursor get_key] */
		ret = cursor->get_key(cursor, &lsn.file, &lsn.offset,
		    &opcount);
		/*! [log cursor get_key] */
		/*
		 * Save one of the LSNs we get back to search for it
		 * later.  Pick a later one because we want to walk from
		 * that LSN to the end (where the multi-step transaction
		 * was performed).  Just choose the record that is MAX_KEYS.
		 */
		if (++i == MAX_KEYS)
			lsnsave = lsn;
		/*! [log cursor get_value] */
		ret = cursor->get_value(cursor, &txnid, &rectype,
		    &optype, &fileid, &logrec_key, &logrec_value);
		/*! [log cursor get_value] */
		RECORD_PRINT;
		/*
		 * If we are in a transaction and this is a new one, end
		 * the previous one.
		 */
		if (in_txn && opcount == 0) {
			ret = session2->commit_transaction(session2, NULL);
			in_txn = 0;
		}

		/*
		 * If the operation is a put, replay it here on the backup
		 * connection.  Note, we cheat by looking only for fileid 1
		 * in this example.  The metadata is fileid 0.
		 */
		if (fileid == 1 && rectype == WT_LOGREC_COMMIT &&
		    optype == WT_LOGOP_ROW_PUT) {
			if (!in_txn) {
				ret = session2->begin_transaction(session2,
				    NULL);
				in_txn = 1;
			}
			cursor2->set_key(cursor2, &logrec_key);
			cursor2->set_value(cursor2, &logrec_value);
			ret = cursor2->insert(cursor2);
		}
	}
	if (in_txn) {
		ret = session2->commit_transaction(session2, NULL);
		in_txn = 0;
	}
	cursor2->close(cursor2);
	/*
	 * Compare the tables after replay.  They should be identical.
	 */
	if (compare_tables(session, session2))
		printf("compare failed\n");
	session2->close(session2, NULL);
	wt_conn2->close(wt_conn2, NULL);

	cursor->reset(cursor);
	/*! [log cursor set_key] */
	cursor->set_key(cursor, lsnsave.file, lsnsave.offset, 0, 0);
	/*! [log cursor set_key] */
	/*! [log cursor search] */
	ret = cursor->search(cursor);
	/*! [log cursor search] */
	printf("Reset to saved ");
	/*
	 * Walk all records starting with this key.
	 */
	first = 1;
	while ((ret = cursor->get_key(cursor,
	    &lsn.file, &lsn.offset, &opcount)) == 0) {
		if (first) {
			first = 0;
			assert(lsnsave.file == lsn.file &&
			    lsnsave.offset == lsn.offset);
		}
		ret = cursor->get_value(cursor, &txnid, &rectype,
		    &optype, &fileid, &logrec_key, &logrec_value);
		RECORD_PRINT;
		ret = cursor->next(cursor);
		if (ret != 0)
			break;
	}
	cursor->close(cursor);
	return (ret);
}

int main(void)
{
	WT_CONNECTION *wt_conn;
	WT_CURSOR *cursor;
	WT_SESSION *session;
	int i, ret;
	char cmd_buf[256], k[16], v[16];

	snprintf(cmd_buf, sizeof(cmd_buf), "rm -rf %s %s && mkdir %s %s",
	    home, home2, home, home2);
	if ((ret = system(cmd_buf)) != 0) {
		fprintf(stderr, "%s: failed ret %d\n", cmd_buf, ret);
		return (ret);
	}
	if ((ret = wiredtiger_open(home, NULL,
	    CONN_CONFIG, &wt_conn)) != 0) {
		fprintf(stderr, "Error connecting to %s: %s\n",
		    home, wiredtiger_strerror(ret));
		return (ret);
	}

	ret = wt_conn->open_session(wt_conn, NULL, NULL, &session);
	ret = session->create(session, uri,
	    "key_format=S,value_format=S");

	ret = session->open_cursor(session, uri, NULL, NULL, &cursor);
	/*
	 * Perform some operations with individual auto-commit transactions.
	 */
	for (i = 0; i < MAX_KEYS; i++) {
		snprintf(k, sizeof(k), "key%d", i);
		snprintf(v, sizeof(v), "value%d", i);
		cursor->set_key(cursor, k);
		cursor->set_value(cursor, v);
		ret = cursor->insert(cursor);
	}
	ret = session->begin_transaction(session, NULL);
	/*
	 * Perform some operations within a single transaction.
	 */
	for (i = MAX_KEYS; i < MAX_KEYS+5; i++) {
		snprintf(k, sizeof(k), "key%d", i);
		snprintf(v, sizeof(v), "value%d", i);
		cursor->set_key(cursor, k);
		cursor->set_value(cursor, v);
		ret = cursor->insert(cursor);
	}
	ret = session->commit_transaction(session, NULL);
	cursor->close(cursor);

	/*
	 * Close and reopen the connection so that the log ends up with
	 * a variety of records such as file sync and checkpoint.  We
	 * have archiving turned off.
	 */
	ret = wt_conn->close(wt_conn, NULL);
	if ((ret = wiredtiger_open(home, NULL,
	    CONN_CONFIG, &wt_conn)) != 0) {
		fprintf(stderr, "Error connecting to %s: %s\n",
		    home, wiredtiger_strerror(ret));
		return (ret);
	}

	ret = wt_conn->open_session(wt_conn, NULL, NULL, &session);
	ret = walk_log(session);
	ret = wt_conn->close(wt_conn, NULL);
	return (ret);
}

File: examples/java/com/wiredtiger/examples/ex_log.java
/*-
 * Public Domain 2014-2016 MongoDB, Inc.
 * Public Domain 2008-2014 WiredTiger, Inc.
 *
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * ex_log.java
 * 	demonstrates how to logging and log cursors.
 */
package com.wiredtiger.examples;
import com.wiredtiger.db.*;
import java.io.*;
import java.util.*;

class Lsn {
    int file;
    long offset;
}

public class ex_log {

    public static String home1 = "WT_HOME_LOG_1";
    public static String home2 = "WT_HOME_LOG_2";
    public static String uri = "table:logtest";

    public static final String CONN_CONFIG =
        "create,cache_size=100MB,log=(archive=false,enabled=true)";
    public static final int MAX_KEYS = 10;

    static Session
    setup_copy()
        throws WiredTigerException
    {
        int ret = 0;
        Connection conn;

        conn = wiredtiger.open(home2, CONN_CONFIG);
        Session session = conn.open_session(null);
        session.create(uri, "key_format=U,value_format=U");
        return (session);
    }

    static int
    compare_tables(Session session, Session sess_copy)
        throws WiredTigerException
    {
        int ret;

        Cursor cursor = session.open_cursor(uri, null, null);
        Cursor curs_copy = sess_copy.open_cursor(uri, null, null);

        while ((ret = cursor.next()) == 0) {
            ret = curs_copy.next();
            byte[] key = cursor.getKeyByteArray();
            byte[] value = cursor.getValueByteArray();
            byte[] key_copy = curs_copy.getKeyByteArray();
            byte[] value_copy = curs_copy.getValueByteArray();
            if (!Arrays.equals(key, key_copy) ||
                !Arrays.equals(value, value_copy)) {
                System.err.println(
                    "Mismatched: key " + new String(key) +
                    ", key_copy " + new String(key_copy) +
                    ", value " + new String(value) +
                    ", value_copy " + new String(value_copy));
                return (1);
            }
        }
        if (ret != wiredtiger.WT_NOTFOUND)
            System.err.println("WT_CURSOR.next: " +
                wiredtiger.wiredtiger_strerror(ret));
        ret = cursor.close();

        ret = curs_copy.next();
        if (ret != wiredtiger.WT_NOTFOUND)
            System.err.println("WT_CURSOR.next: " +
                wiredtiger.wiredtiger_strerror(ret));
        ret = curs_copy.close();

        return (ret);
    }

    /*! [log cursor walk] */
    static void
    print_record(Lsn lsn, int opcount,
       int rectype, int optype, long txnid, int fileid,
       byte[] key, byte[] value)
    {
        System.out.print(
            "LSN [" + lsn.file + "][" + lsn.offset + "]." + opcount +
            ": record type " + rectype + " optype " + optype +
            " txnid " + txnid + " fileid " + fileid);
        System.out.println(" key size " + key.length +
            " value size " + value.length);
        if (rectype == wiredtiger.WT_LOGREC_MESSAGE)
            System.out.println("Application Record: " + new String(value));
    }

    /*
     * simple_walk_log --
     *	A simple walk of the log.
     */
    static int
    simple_walk_log(Session session)
        throws WiredTigerException
    {
        Cursor cursor;
        Lsn lsn = new Lsn();
        byte[] logrec_key, logrec_value;
        long txnid;
        int fileid, opcount, optype, rectype;
        int ret;

        /*! [log cursor open] */
        cursor = session.open_cursor("log:", null, null);
        /*! [log cursor open] */

        while ((ret = cursor.next()) == 0) {
            /*! [log cursor get_key] */
            lsn.file = cursor.getKeyInt();
            lsn.offset = cursor.getKeyLong();
            opcount = cursor.getKeyInt();
            /*! [log cursor get_key] */
            /*! [log cursor get_value] */
            txnid = cursor.getValueLong();
            rectype = cursor.getValueInt();
            optype = cursor.getValueInt();
            fileid = cursor.getValueInt();
            logrec_key = cursor.getValueByteArray();
            logrec_value = cursor.getValueByteArray();
            /*! [log cursor get_value] */

            print_record(lsn, opcount,
                rectype, optype, txnid, fileid, logrec_key, logrec_value);
        }
        if (ret == wiredtiger.WT_NOTFOUND)
            ret = 0;
        ret = cursor.close();
        return (ret);
    }
    /*! [log cursor walk] */

    static int
    walk_log(Session session)
        throws WiredTigerException
    {
        Connection wt_conn2;
        Cursor cursor, cursor2;
        Lsn lsn, lsnsave;
        byte[] logrec_key, logrec_value;
        Session session2;
        long txnid;
        int fileid, opcount, optype, rectype;
        int i, ret;
        boolean in_txn, first;

        session2 = setup_copy();
        wt_conn2 = session2.getConnection();
        cursor = session.open_cursor("log:", null, null);
        cursor2 = session2.open_cursor(uri, null, "raw=true");
        i = 0;
        in_txn = false;
        txnid = 0;
        lsn = new Lsn();
        lsnsave = new Lsn();
        while ((ret = cursor.next()) == 0) {
            lsn.file = cursor.getKeyInt();
            lsn.offset = cursor.getKeyLong();
            opcount = cursor.getKeyInt();

            /*
             * Save one of the LSNs we get back to search for it
             * later.  Pick a later one because we want to walk from
             * that LSN to the end (where the multi-step transaction
             * was performed).  Just choose the record that is MAX_KEYS.
             */
            if (++i == MAX_KEYS)
                lsnsave = lsn;
            txnid = cursor.getValueLong();
            rectype = cursor.getValueInt();
            optype = cursor.getValueInt();
            fileid = cursor.getValueInt();
            logrec_key = cursor.getValueByteArray();
            logrec_value = cursor.getValueByteArray();

            print_record(lsn, opcount,
                rectype, optype, txnid, fileid, logrec_key, logrec_value);

            /*
             * If we are in a transaction and this is a new one, end
             * the previous one.
             */
            if (in_txn && opcount == 0) {
                ret = session2.commit_transaction(null);
                in_txn = false;
            }

            /*
             * If the operation is a put, replay it here on the backup
             * connection.  Note, we cheat by looking only for fileid 1
             * in this example.  The metadata is fileid 0.
             */
            if (fileid == 1 && rectype == wiredtiger.WT_LOGREC_COMMIT &&
                optype == wiredtiger.WT_LOGOP_ROW_PUT) {
                if (!in_txn) {
                    ret = session2.begin_transaction(null);
                    in_txn = true;
                }
                cursor2.putKeyByteArray(logrec_key);
                cursor2.putValueByteArray(logrec_value);
                ret = cursor2.insert();
            }
        }
        if (in_txn)
            ret = session2.commit_transaction(null);

        ret = cursor2.close();
        /*
         * Compare the tables after replay.  They should be identical.
         */
        if (compare_tables(session, session2) != 0)
            System.out.println("compare failed");
        ret = session2.close(null);
        ret = wt_conn2.close(null);

        ret = cursor.reset();
        /*! [log cursor set_key] */
        cursor.putKeyInt(lsnsave.file);
        cursor.putKeyLong(lsnsave.offset);
        /*! [log cursor set_key] */
        /*! [log cursor search] */
        ret = cursor.search();
        /*! [log cursor search] */
        System.out.println("Reset to saved...");
        /*
         * Walk all records starting with this key.
         */
        first = true;
        while (ret == 0) {  /*TODO: not quite right*/
            lsn.file = cursor.getKeyInt();
            lsn.offset = cursor.getKeyLong();
            opcount = cursor.getKeyInt();
            if (first) {
                first = false;
                if (lsnsave.file != lsn.file ||
                    lsnsave.offset != lsn.offset) {
                    System.err.println("search returned the wrong LSN");
                    System.exit(1);
                }
            }
            txnid = cursor.getValueLong();
            rectype = cursor.getValueInt();
            optype = cursor.getValueInt();
            fileid = cursor.getValueInt();
            logrec_key = cursor.getValueByteArray();
            logrec_value = cursor.getValueByteArray();

            print_record(lsn, opcount, rectype, optype, txnid,
                fileid, logrec_key, logrec_value);

            ret = cursor.next();
            if (ret != 0)
                break;
        }
        ret = cursor.close();
        return (ret);
    }

    public static int
    logExample()
        throws WiredTigerException
    {
        Connection wt_conn;
        Cursor cursor;
        Session session;
        int i, record_count, ret;

        try {
            String command = "/bin/rm -rf " + home1 + " " + home2;
            Process proc = Runtime.getRuntime().exec(command);
            BufferedReader br = new BufferedReader(
                new InputStreamReader(proc.getInputStream()));
            while(br.ready())
                System.out.println(br.readLine());
            br.close();
            proc.waitFor();
            new File(home1).mkdir();
            new File(home2).mkdir();
        } catch (Exception ex) {
            System.err.println("Exception: " + ex);
            return (1);
        }
        if ((wt_conn = wiredtiger.open(home1, CONN_CONFIG)) == null) {
            System.err.println("Error connecting to " + home1);
            return (1);
        }

        session = wt_conn.open_session(null);
        ret = session.create(uri, "key_format=S,value_format=S");

        cursor = session.open_cursor(uri, null, null);
        /*
         * Perform some operations with individual auto-commit transactions.
         */
        for (record_count = 0, i = 0; i < MAX_KEYS; i++, record_count++) {
            String k = "key" + i;
            String v = "value" + i;
            cursor.putKeyString(k);
            cursor.putValueString(v);
            ret = cursor.insert();
        }
        ret = session.begin_transaction(null);
        /*
         * Perform some operations within a single transaction.
         */
        for (i = MAX_KEYS; i < MAX_KEYS+5; i++, record_count++) {
            String k = "key" + i;
            String v = "value" + i;
            cursor.putKeyString(k);
            cursor.putValueString(v);
            ret = cursor.insert();
        }
        ret = session.commit_transaction(null);
        ret = cursor.close();

        /*! [log cursor printf] */
        ret = session.log_printf("Wrote " + record_count + " records");
        /*! [log cursor printf] */

        session.close(null);
        /*
         * Close and reopen the connection so that the log ends up with
         * a variety of records such as file sync and checkpoint.  We
         * have archiving turned off.
         */
        ret = wt_conn.close(null);
        if ((wt_conn = wiredtiger.open(home1, CONN_CONFIG)) == null) {
            System.err.println("Error connecting to " + home1);
            return (ret);
        }

        session = wt_conn.open_session(null);
        ret = simple_walk_log(session);
        ret = walk_log(session);
        ret = session.close(null);
        ret = wt_conn.close(null);
        return (ret);
    }

    public static void
    main(String[] args)
    {
        try {
            System.exit(logExample());
        }
        catch (WiredTigerException wte) {
            System.err.println("Exception: " + wte);
            wte.printStackTrace();
            System.exit(1);
        }
    }
}
